### PR TITLE
feat: SecretProvider for storing/retrieving secrets

### PR DESCRIPTION
**Pull request by lenny-intel**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**If your build fails** due to your commit message not passing the build checks, please review the guidelines here: https://github.com/edgexfoundry/device-sdk-c/blob/master/.github/CONTRIBUTING.md

## What is the new behavior?
Added creation and initialization SecretProvider during bootstrap for both secure and non-secure modes
Added InsecureSecrets and SecretStore config to device-simple as samples and used to test initializing SecretProvider in secure mode.
APIs to take advantage of the SecretProvider  are TBD in future PRs

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

## New Imports

- [ ] Yes
- [x] No

## Specific Instructions

This PR is about just about creating and initializing the Secret Provider. Using it come in following PR.
**Do the following to test this PR:**

- clone latest developer scripts repo
- Modify developer-scripts/releases/nightly-build/compose-files/docker-compose-nexus.yml as follows:
  - Add `ADD_SECRETSTORE_TOKENS: "device-simple"` to `environment` section for `vault-worker` 
 
```
    environment:
      SECRETSTORE_SETUP_DONE_FLAG: /tmp/edgex/secrets/edgex-consul/.secretstore-setup-done
      ADD_SECRETSTORE_TOKENS: "device-simple"
    hostname: edgex-vault-worker
```

- Run compose from developer-scripts/releases/nightly-build/compose-files
  - `make run`
- build Device SDK from branch for this PR (i.e. clone the branch)

**Run Device Simple in secure mode**
- cd to `example/cmd/device-simple/`
- sudo EDGEX_SECURITY_SECRET_STORE=true ./device-simple
  - Need to run as root to access vault token file
- Verify service starts successfully with the following in the log messages:

```
level=INFO ts=2021-01-08T21:44:12.7857283Z app=device-simple source=secret.go:51 msg="Creating SecretClient"
level=INFO ts=2021-01-08T21:44:12.7872455Z app=device-simple source=secret.go:58 msg="Reading secret store configuration and authentication token"
level=INFO ts=2021-01-08T21:44:12.7873587Z app=device-simple source=secret.go:70 msg="Attempting to create secret client"
level=INFO ts=2021-01-08T21:44:12.805743Z app=device-simple source=secret.go:75 msg="Created SecretClient"
```

**Run Device Simple in non-secure mode**
- EDGEX_SECURITY_SECRET_STORE=false ./device-simple
- Verify service starts successfully without any log message about SecretClient

### CODE REVIEW

File: internal/v2/application/command.go [hunk 1]
```diff
@@ -491,28 +493,28 @@ func createCommandValueFromDeviceResource(dr *contract.DeviceResource, v string)
 
 	origin := time.Now().UnixNano()
 	switch strings.ToLower(dr.Properties.Value.Type) {
-	case strings.ToLower(dtos.ValueTypeString):
+	case strings.ToLower(v2.ValueTypeString):
```
Comment (lenny-intel) [inline internal/v2/application/command.go:496]: These un-related changes cause by latest go-mod-core-contracts.

File: internal/autoevent/executor_test.go [hunk 2]
```diff
@@ -20,7 +20,7 @@ func TestCompareReadings(t *testing.T) {
 	readings[2] = contract.Reading{Name: "Pressure", Value: "3"}
 	readings[3] = contract.Reading{Name: "Image", BinaryValue: []byte("This is a image")}
 
-	lc := logger.NewClientStdOut("device-sdk-test", false, "DEBUG")
+	lc := logger.NewMockClient()
```
Comment (lenny-intel) [inline internal/autoevent/executor_test.go:23]: Unrelated change due to latest core contracts.

**Review by hahattan — DISMISSED** [review]
LGTM

**Review by hahattan — APPROVED** [review]
LGTM

File: pkg/service/main.go [hunk 3]
```diff
@@ -59,17 +60,18 @@ func Main(serviceName string, serviceVersion string, proto interface{}, ctx cont
...
 			autodiscovery.BootstrapHandler,
-			message.NewBootstrap(serviceName, serviceVersion).BootstrapHandler,
+			handlers.NewStartMessage(serviceName, serviceVersion).BootstrapHandler,
 		})
 
 	ds.Stop(false)
 }
 
-func setServiceName(name string, profile string) string {
+func setServiceName(name string, _ string) string {
```
Comment (tonyespy) [inline pkg/service/main.go:74]: Since this is an internal function and this parameter clearly isn't being used anymore, wouldn't it be better to just remove it altogether vs. using a blank identifier?

Also as this is seemingly unrelated to the SecretProvider changes, IMHOP it belongs in its own commit.

File: pkg/service/main.go [hunk 4]
```diff
@@ -14,19 +14,20 @@ import (
...
 	"github.com/edgexfoundry/go-mod-bootstrap/bootstrap/startup"
 	"github.com/edgexfoundry/go-mod-bootstrap/di"
+
 	"github.com/gorilla/mux"
 )
 
 var instanceName string
 
-func Main(serviceName string, serviceVersion string, proto interface{}, ctx context.Context, cancel context.CancelFunc, router *mux.Router, readyStream chan<- bool) {
+func Main(serviceName string, serviceVersion string, proto interface{}, ctx context.Context, cancel context.CancelFunc, router *mux.Router, _ chan<- bool) {
```
Comment (tonyespy) [inline pkg/service/main.go:30]: Why change final `readyStream` parameter to the blank identifier? If it's not being used, why not just remove it? Also as this is unrelated to the SecretProvider changes, IMHOP it belongs in a separate cleanup/re-factor commit.

One other question that should maybe be discussed at next week's device service WG relates to whether or not this should package should still even remain public...

File: go.mod [hunk 5]
```diff
@@ -3,13 +3,13 @@ module github.com/edgexfoundry/device-sdk-go
...
-	github.com/edgexfoundry/go-mod-core-contracts v0.1.115
-	github.com/edgexfoundry/go-mod-registry v0.1.26
+	github.com/edgexfoundry/go-mod-bootstrap v0.0.68
+	github.com/edgexfoundry/go-mod-core-contracts v0.1.135
+	github.com/edgexfoundry/go-mod-registry v0.1.27
 	github.com/fxamacker/cbor/v2 v2.2.0
 	github.com/google/uuid v1.1.2
 	github.com/gorilla/mux v1.8.0
-	github.com/stretchr/testify v1.5.1
+	github.com/stretchr/testify v1.6.1
```
Comment (tonyespy) [inline go.mod:12]: Is there some reason the testify version is being bumped here?

**Review by tonyespy — COMMENTED** [review]
In general this looks pretty good, although I had some questions about what appear to be some unrelated changes.

Also I wasn't sure how to test this PR. Could you provide some basic instructions on how to do so?

File: internal/v2/application/command.go (hunk 1 above)
Comment (lenny-intel) [reply internal/v2/application/command.go:496]: IMO, minor names changes don't justify it when can be handle here. Would agree if required logic changes.

File: internal/common/types.go [hunk 6]
```diff
@@ -57,7 +58,7 @@ type ServiceInfo struct {
 // DeviceInfo is a struct which contains device specific configuration settings.
 type DeviceInfo struct {
 	// DataTransform specifies whether or not the DS perform transformations
-	// specified by valuedescriptor on a actuation or query command.
+	// specified by value descriptor on a actuation or query command.
```
[the remark this answers is not in the document]
Comment (lenny-intel) [reply internal/common/types.go:61]: Don't think the SDK has been update to reflect that yet.

File: internal/common/types.go [hunk 7]
```diff
@@ -68,15 +69,15 @@ type DeviceInfo struct {
...
+	// result (including the value descriptor name) that can be returned
 	// by a Driver.
 	MaxCmdValueLen int
 	// InitCmd specifies a device resource command which is automatically
 	// generated whenever a new device is removed from the DS.
 	RemoveCmd string
 	// RemoveCmdArgs specify arguments to be used when building the RemoveCmd.
 	RemoveCmdArgs string
-	// ProfilesDir specifies a directory which contains deviceprofile
+	// ProfilesDir specifies a directory which contains device profile
```
[the remark this answers is not in the document]
Comment (lenny-intel) [reply internal/common/types.go:80]: Sure.

File: internal/v2/application/command.go [hunk 8]
```diff
@@ -23,12 +23,14 @@ import (
...
 	"github.com/edgexfoundry/device-sdk-go/internal/transformer"
 	dsModels "github.com/edgexfoundry/device-sdk-go/pkg/models"
+
 	bootstrapContainer "github.com/edgexfoundry/go-mod-bootstrap/bootstrap/container"
 	"github.com/edgexfoundry/go-mod-bootstrap/di"
 	"github.com/edgexfoundry/go-mod-core-contracts/clients/coredata"
 	"github.com/edgexfoundry/go-mod-core-contracts/clients/logger"
 	edgexErr "github.com/edgexfoundry/go-mod-core-contracts/errors"
 	contract "github.com/edgexfoundry/go-mod-core-contracts/models"
+	v2 "github.com/edgexfoundry/go-mod-core-contracts/v2"
```
[the remark this answers is not in the document]
Comment (lenny-intel) [reply internal/v2/application/command.go:33]: Not sure why the IDE added the alias. Will fix.

File: pkg/service/main.go (hunk 3 above)
Comment (lenny-intel) [reply pkg/service/main.go:74]: I was just fixing minor issues called out by the IDE. I agree refactoring it to remove is probably a good idea for another PR.

File: internal/v2/application/command.go [hunk 9]
```diff
@@ -717,13 +719,10 @@ func commandValueToReading(cv *dsModels.CommandValue, deviceName string, mediaTy
 		encoding = dsModels.DefaultFloatEncoding
 	}
 
-	reading := &dtos.BaseReading{Name: cv.DeviceResourceName, DeviceName: deviceName, ValueType: cv.ValueTypeToString()}
+	reading := &dtos.BaseReading{ResourceName: cv.DeviceResourceName, DeviceName: deviceName, ValueType: cv.ValueTypeToString()}
 	if cv.Type == dsModels.Binary {
 		reading.BinaryValue = cv.BinValue
 		reading.MediaType = mediaType
-	} else if cv.Type == dsModels.Float32 || cv.Type == dsModels.Float64 {
```
[the remark this answers is not in the document]
Comment (lenny-intel) [reply internal/v2/application/command.go:724]: Yes.

File: pkg/service/main.go (hunk 4 above)
Comment (lenny-intel) [reply pkg/service/main.go:30]: Fix another minor issue that IDE flagged. I agree with the rest of your comments.

File: go.mod (hunk 5 above)
Comment (lenny-intel) [reply go.mod:12]: Yes, because we are now using v1.6.1 in all repos.

**Review by cloudxxx8 — APPROVED** [review]
LGTM

**Comment by tonyespy** [comment]
@cloudxxx8 why did you merge this? Per my last review I still had some testing to do and hadn't yet approved this.

**Comment by tonyespy** [comment]
@lenny-intel I re-tested today using the edgexfoundry snap. I added the following override to security-secretstore-setup (which in turn execs security-file-token-provider) and the token is generated for device-simple. I then manually copied the file from the `$SNAP_DATA/secrets/device-simple` to the `/res` directory of device-simple, updated the `configuration.toml` path for the token and everything worked. I also verified that device-simple works with `EDGEX_SECURITY_SECRET_STORE=false` as well.

Note, when I tried testing by adding `ADD_SECRETSTORE_TOKENS` to vault-worked in docker-compose-nexus.yml per your instructions, the device-simple token was generated for me when I brought everything up.

**Comment by cloudxxx8** [comment]
sorry, @tonyespy , I should have read through your comments thoroughly.  I merged this PR because I saw you put comments not "change request" and Lenny had resolved all the comments.  Also, I would like to unblock other two PRs early.
